File: src/main/java/org/moinex/ui/dialog/investment/AddTickerController.java
/*
 * Filename: AddTickerController.java
 * Created on: January  8, 2025
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 */

package org.moinex.ui.dialog.investment;

import com.jfoenix.controls.JFXButton;
import jakarta.persistence.EntityExistsException;
import java.math.BigDecimal;
import javafx.fxml.FXML;
import javafx.stage.Stage;
import lombok.NoArgsConstructor;
import org.moinex.model.enums.AssetType;
import org.moinex.service.PreferencesService;
import org.moinex.service.TickerService;
import org.moinex.util.Constants;
import org.moinex.util.UIUtils;
import org.moinex.util.WindowUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Controller;

/** Controller for the Add Ticker dialog */
@Controller
@NoArgsConstructor
public final class AddTickerController extends BaseTickerManagement {
    @FXML private JFXButton yahooLookupButton;

    /**
     * Constructor
     *
     * @param tickerService Ticker service
     * @param preferencesService I18n service
     * @note This constructor is used for dependency injection
     */
    @Autowired
    public AddTickerController(TickerService tickerService, PreferencesService preferencesService) {
        super(tickerService, preferencesService);
    }

    @FXML
    @Override
    protected void initialize() {
        super.initialize();

        UIUtils.addTooltipToNode(
                yahooLookupButton,
                preferencesService.translate(
                        Constants.TranslationKeys.INVESTMENT_BUTTON_YAHOO_LOOKUP_TOOLTIP));
    }

    @FXML
    @Override
    protected void handleSave() {
        String name = nameField.getText();
        String symbol = symbolField.getText();
        String currentPriceStr = currentPriceField.getText();
        AssetType type = typeComboBox.getValue();
        String quantityStr = quantityField.getText();
        String avgUnitPriceStr = avgUnitPriceField.getText();

        if (name == null
                || symbol == null
                || currentPriceStr == null
                || type == null
                || name.isBlank()
                || symbol.isBlank()
                || currentPriceStr.isBlank()) {
            WindowUtils.showInformationDialog(
                    preferencesService.translate(
                            Constants.TranslationKeys.INVESTMENT_DIALOG_EMPTY_FIELDS_TITLE),
                    preferencesService.translate(
                            Constants.TranslationKeys.INVESTMENT_DIALOG_EMPTY_FIELDS_MESSAGE));

            return;
        }

        // If quantity is set, then avgUnitPrice must be set or vice versa
        if ((quantityStr == null || quantityStr.isBlank())
                        && !(avgUnitPriceStr == null || avgUnitPriceStr.isBlank())
                || (avgUnitPriceStr == null || avgUnitPriceStr.isBlank())
                        && !(quantityStr == null || quantityStr.isBlank())) {
            WindowUtils.showInformationDialog(
                    preferencesService.translate(
                            Constants.TranslationKeys.INVESTMENT_DIALOG_INVALID_FIELDS_TITLE),
                    preferencesService.translate(
                            Constants.TranslationKeys.INVESTMENT_DIALOG_INVALID_FIELDS_MESSAGE));

            return;
        }

        try {
            BigDecimal currentPrice = new BigDecimal(currentPriceStr);

            BigDecimal quantity;
            BigDecimal avgUnitPrice;

            if ((quantityStr == null || quantityStr.isBlank())
                    && (avgUnitPriceStr == null || avgUnitPriceStr.isBlank())) {
                quantity = BigDecimal.ZERO;
                avgUnitPrice = BigDecimal.ZERO;
            } else {
                assert quantityStr != null;
                quantity = new BigDecimal(quantityStr);
                assert avgUnitPriceStr != null;
                avgUnitPrice = new BigDecimal(avgUnitPriceStr);
            }

            tickerService.addTicker(name, symbol, type, currentPrice, avgUnitPrice, quantity);

            WindowUtils.showSuccessDialog(
                    preferencesService.translate(
                            Constants.TranslationKeys.INVESTMENT_DIALOG_TICKER_ADDED_TITLE),
                    preferencesService.translate(
                            Constants.TranslationKeys.INVESTMENT_DIALOG_TICKER_ADDED_MESSAGE));

            Stage stage = (Stage) nameField.getScene().getWindow();
            stage.close();
        } catch (NumberFormatException e) {
            WindowUtils.showErrorDialog(
                    preferencesService.translate(
                            Constants.TranslationKeys.INVESTMENT_DIALOG_INVALID_NUMBER_TITLE),
                    preferencesService.translate(
                            Constants.TranslationKeys.INVESTMENT_DIALOG_INVALID_NUMBER_MESSAGE));
        } catch (IllegalArgumentException | EntityExistsException e) {
            WindowUtils.showErrorDialog(
                    preferencesService.translate(
                            Constants.TranslationKeys.INVESTMENT_DIALOG_ERROR_ADDING_TICKER_TITLE),
                    e.getMessage());
        }
    }

    @FXML
    private void goToYahooLookup() {
        WindowUtils.openUrl(Constants.YAHOO_LOOKUP_URL);
    }
}

File: src/main/kotlin/org/moinex/ui/dialog/investment/create/AddTickerController.kt
/*
 * Filename: AddTickerController.kt (original filename: AddTickerController.java)
 * Created on: January  8, 2025
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 *
 * Migrate to Kotlin on 20/03/2026
 */

package org.moinex.ui.dialog.investment.create

import com.jfoenix.controls.JFXButton
import jakarta.persistence.EntityExistsException
import javafx.fxml.FXML
import javafx.stage.Stage
import org.moinex.common.constant.Constants
import org.moinex.common.constant.TranslationKeys
import org.moinex.common.util.UIUtils
import org.moinex.common.util.WindowUtils
import org.moinex.model.investment.Ticker
import org.moinex.service.PreferencesService
import org.moinex.service.investment.TickerService
import org.moinex.ui.dialog.investment.base.BaseTickerManagement
import org.springframework.stereotype.Controller
import java.math.BigDecimal
import java.time.LocalDateTime

@Controller
class AddTickerController(
    tickerService: TickerService,
    preferencesService: PreferencesService,
) : BaseTickerManagement(tickerService, preferencesService) {
    @FXML
    private lateinit var yahooLookupButton: JFXButton

    @FXML
    override fun initialize() {
        super.initialize()

        UIUtils.addTooltipToNode(
            yahooLookupButton,
            preferencesService.translate(TranslationKeys.INVESTMENT_BUTTON_YAHOO_LOOKUP_TOOLTIP),
        )
    }

    @FXML
    override fun handleSave() {
        val name = nameField.text
        val symbol = symbolField.text
        val currentPriceStr = currentPriceField.text
        val type = typeComboBox.value
        val quantityStr = quantityField.text
        val avgUnitPriceStr = avgUnitPriceField.text

        if (name.isNullOrBlank() || symbol.isNullOrBlank() || currentPriceStr.isNullOrBlank() || type == null) {
            WindowUtils.showInformationDialog(
                preferencesService.translate(TranslationKeys.INVESTMENT_DIALOG_EMPTY_FIELDS_TITLE),
                preferencesService.translate(TranslationKeys.INVESTMENT_DIALOG_EMPTY_FIELDS_MESSAGE),
            )
            return
        }

        if (quantityStr.isNullOrBlank() xor avgUnitPriceStr.isNullOrBlank()) {
            WindowUtils.showInformationDialog(
                preferencesService.translate(TranslationKeys.INVESTMENT_DIALOG_INVALID_FIELDS_TITLE),
                preferencesService.translate(TranslationKeys.INVESTMENT_DIALOG_INVALID_FIELDS_MESSAGE),
            )
            return
        }

        runCatching {
            val currentPrice = currentPriceStr.toBigDecimal()

            val (quantity, avgUnitPrice) =
                if (quantityStr.isNullOrBlank() && avgUnitPriceStr.isNullOrBlank()) {
                    BigDecimal.ZERO to BigDecimal.ZERO
                } else {
                    quantityStr!!.toBigDecimal() to avgUnitPriceStr!!.toBigDecimal()
                }

            tickerService.createTicker(
                Ticker(
                    createdAt = LocalDateTime.now(),
                    lastUpdate = LocalDateTime.now(),
                    name = name,
                    symbol = symbol,
                    type = type,
                    currentQuantity = quantity,
                    currentUnitValue = currentPrice,
                    averageUnitValue = avgUnitPrice,
                ),
            )

            WindowUtils.showSuccessDialog(
                preferencesService.translate(TranslationKeys.INVESTMENT_DIALOG_TICKER_ADDED_TITLE),
                preferencesService.translate(TranslationKeys.INVESTMENT_DIALOG_TICKER_ADDED_MESSAGE),
            )

            (nameField.scene.window as Stage).close()
        }.onFailure { e ->
            when (e) {
                is NumberFormatException -> {
                    WindowUtils.showErrorDialog(
                        preferencesService.translate(TranslationKeys.INVESTMENT_DIALOG_INVALID_NUMBER_TITLE),
                        preferencesService.translate(TranslationKeys.INVESTMENT_DIALOG_INVALID_NUMBER_MESSAGE),
                    )
                }
                is IllegalArgumentException, is EntityExistsException -> {
                    WindowUtils.showErrorDialog(
                        preferencesService.translate(TranslationKeys.INVESTMENT_DIALOG_ERROR_ADDING_TICKER_TITLE),
                        e.message ?: "",
                    )
                }
                else -> throw e
            }
        }
    }

    @FXML
    private fun goToYahooLookup() {
        WindowUtils.openUrl(Constants.YAHOO_LOOKUP_URL)
    }
}
